src/reader/metadata.rs

```diff
@@ -22,153 +47,234 @@ pub struct FileMetadata {
-                stripe_footers.push(deserialize_stripe_footer(
-                    &scratch,
-                    compression,
-                )?);
-            }
-
-            Ok(FileMetadata {
-                postscript,
-                footer,
-                metadata,
-                stripe_footers,
-            })
-        }
+    // The final byte of the file contains the serialized length of the Postscript,
+    // which must be less than 256 bytes.
+    let postscript_len = tail_bytes[tail_bytes.len() - 1] as u64;
+    tail_bytes.truncate(tail_bytes.len() - 1);
+
+    // TODO: slice here could panic if file too small
+    let postscript = PostScript::decode(&tail_bytes[tail_bytes.len() - postscript_len as usize..])
+        .context(error::DecodeProtoSnafu)?;
+    let compression =
+        Compression::from_proto(postscript.compression(), postscript.compression_block_size);
+    tail_bytes.truncate(tail_bytes.len() - postscript_len as usize);
+
+    let footer_length = postscript.footer_length.context(error::OutOfSpecSnafu {
+        msg: "Footer length is empty",
+    })?;
+    let metadata_length = postscript.metadata_length.context(error::OutOfSpecSnafu {
+        msg: "Metadata length is empty",
+    })?;
+
+    // Ensure we have enough bytes for Footer and Metadata
+    let mut tail_bytes = if footer_length + metadata_length > tail_bytes.len() as u64 {
+        // Need second read
+        // -1 is the postscript length byte
+        let offset = file_len - 1 - postscript_len - footer_length - metadata_length;
+        let bytes_to_read = (footer_length + metadata_length) - tail_bytes.len() as u64;
+        let mut prepend_bytes = reader
+            .get_bytes(offset, bytes_to_read)
+            .context(error::IoSnafu)?;
+        prepend_bytes.extend(tail_bytes);
+        prepend_bytes
+    } else {
+        tail_bytes
     };
-}
 
-pub fn read_metadata<R>(reader: &mut R) -> Result<FileMetadata>
-where
-    R: Read + Seek,
-{
-    impl_read_metadata!(reader)
+    let footer = deserialize_footer(
+        &tail_bytes[tail_bytes.len() - footer_length as usize..],
+        compression,
+    )?;
+    tail_bytes.truncate(tail_bytes.len() - footer_length as usize);
+
+    let metadata = deserialize_footer_metadata(
+        &tail_bytes[tail_bytes.len() - metadata_length as usize..],
+        compression,
+    )?;
+
+    let mut stripe_footers = Vec::with_capacity(footer.stripes.len());
+
+    // clippy read_zero_byte_vec lint causing issues so init to non-zero length
+    let mut scratch = vec![0];
+    for stripe in &footer.stripes {
+        let offset = stripe.offset() + stripe.index_length() + stripe.data_length();
+        let len = stripe.footer_length();
+
+        let mut read = reader.get_read(offset).context(error::IoSnafu)?;
+        scratch.resize(len as usize, 0);
+        read.read_exact(&mut scratch).context(error::IoSnafu)?;
+        stripe_footers.push(deserialize_stripe_footer(&scratch, compression)?);
+    }
```

Comment:
Might refactor this away later, we'll see (might not want to decode all stripe footers here and now)